kedro/io/data_catalog.py

```diff
@@ -730,16 +730,22 @@ def list(self, regex_search: str | None = None) -> list[str]:
             ) from exc
         return [dset_name for dset_name in self._datasets if pattern.search(dset_name)]
 
-    def shallow_copy(self) -> DataCatalog:
+    def shallow_copy(
+        self, extra_dataset_patterns: Patterns | None = None
+    ) -> DataCatalog:
         """Returns a shallow copy of the current object.
 
         Returns:
             Copy of the current object.
         """
+        if extra_dataset_patterns:
+            dataset_patterns = {**self._dataset_patterns, **extra_dataset_patterns}
```

Comment:
Could we look at preventing over-writing if users have specified their own default pattern in their catalog?  Like a left join mechanism of sorts